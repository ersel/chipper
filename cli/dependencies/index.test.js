const fs = require('fs');

const dependenciesAction = require('./');
const scanner = require('../../lib/scanner/cache/scanCache');
const openFile = require('../utils/openFile/');

jest.mock('../utils/openFile/');
jest.mock('fs');
jest.mock('../../lib/scanner/cache/scanCache', () =>
	jest.fn(() => {
		const PROJECT_ROOT = process.cwd();
		const MOCK_SCAN_DATA = [
			{
				sourceFile: `${PROJECT_ROOT}/src/documentsService.js`,
				importedModules: [
					{
						source: 'lodash',
						absolute: `${PROJECT_ROOT}/node_modules/lodash/lodash.js`,
						imports: [
							{
								imported: 'uniq',
								local: 'uniq'
							}
						]
					}
				]
			},
			{
				sourceFile: `${PROJECT_ROOT}/src/loginService.js`,
				importedModules: [
					{
						source: './documentsService',
						absolute: `${PROJECT_ROOT}/src/documentsService.js`,
						imports: [
							{
								imported: 'getDocuments',
								local: 'getLoggedInUsersDocuments'
							}
						]
					},
					{
						source: './utils/cookiesService',
						absolute: `${PROJECT_ROOT}/src/utils/cookiesService.js`,
						imports: [
							{
								imported: 'setCookie',
								local: 'setCookie'
							}
						]
					}
				]
			},
			{
				sourceFile: `${PROJECT_ROOT}/src/utils/cookiesServices.js`,
				importedModules: [
					{
						source: 'cookie-jar',
						absolute: `${PROJECT_ROOT}/node_modules/cookie-jar/index.js`,
						imports: [
							{
								imported: 'setCookie',
								local: 'setCookie'
							},
							{
								imported: 'getCookie',
								local: 'getCookie'
							}
						]
					},
					{
						source: 'lodash',
						absolute: `${PROJECT_ROOT}/node_modules/lodash/get/index.js`,
						imports: [
							{
								imported: 'get',
								local: 'get'
							}
						]
					}
				]
			}
		];
		return Promise.resolve(MOCK_SCAN_DATA);
	})
);

describe('dependencies command', () => {
	let testOptions = {};
	beforeEach(() => {
		testOptions = {
			targetDir: '.',
			projectRoot: process.cwd(),
			ext: ['js'],
			incl: ['**/*.js'],
			excl: ['node_modules/**'],
			alias: 'some-alias:/path/to/some-alias',
			fileScanParallelism: 50,
			rescan: true
		};

		openFile.mockReset();
		fs.writeFileSync.mockReset();
	});
	it('should call scanner with the right params', done => {
		dependenciesAction({ target: 'src/' }, testOptions).then(() => {
			expect(scanner).lastCalledWith(
				{
					aliases: { 'some-alias': '/path/to/some-alias' },
					excludedPatterns: ['node_modules/**'],
					extensions: ['js'],
					fileScanParallelism: 50,
					includedPatterns: ['**/*.js'],
					projectRootPath: process.cwd(),
					targetDirectory: '.'
				},
				testOptions.rescan
			);
			done();
		});
	});

	it('should call scanner with the right params when no aliases are defined', done => {
		delete testOptions.alias;
		dependenciesAction({ target: 'src/' }, testOptions).then(() => {
			expect(scanner).lastCalledWith(
				{
					aliases: {},
					excludedPatterns: ['node_modules/**'],
					extensions: ['js'],
					fileScanParallelism: 50,
					includedPatterns: ['**/*.js'],
					projectRootPath: process.cwd(),
					targetDirectory: '.'
				},
				testOptions.rescan
			);
			done();
		});
	});

	it('should return all imports of a single module (absolute path)', done => {
		dependenciesAction(
			{ target: `${process.cwd()}/src/loginService.js` },
			testOptions
		).then(() => {
			expect(fs.writeFileSync.mock.calls).toHaveLength(1);
			const lastReportWriteCall = fs.writeFileSync.mock.calls[0];
			const [fileName, reportContents] = lastReportWriteCall;
			expect(reportContents).toMatchSnapshot();
			expect(openFile).lastCalledWith(fileName);
			expect(openFile).toHaveBeenCalledTimes(1);
			done();
		});
	});

	it('should return all imports of a single module', done => {
		dependenciesAction({ target: 'src/loginService.js' }, testOptions).then(
			() => {
				expect(fs.writeFileSync.mock.calls).toHaveLength(1);
				const lastReportWriteCall = fs.writeFileSync.mock.calls[0];
				const [fileName, reportContents] = lastReportWriteCall;
				expect(reportContents).toMatchSnapshot();
				expect(openFile).lastCalledWith(fileName);
				expect(openFile).toHaveBeenCalledTimes(1);
				done();
			}
		);
	});

	it('should return all imports of a directory of modules', done => {
		dependenciesAction({ target: 'src/' }, testOptions).then(() => {
			expect(fs.writeFileSync.mock.calls).toHaveLength(1);
			const lastReportWriteCall = fs.writeFileSync.mock.calls[0];
			const [fileName, reportContents] = lastReportWriteCall;
			expect(reportContents).toMatchSnapshot();
			expect(openFile).lastCalledWith(fileName);
			expect(openFile).toHaveBeenCalledTimes(1);
			done();
		});
	});

	it('should return all imports for an aliased search target', done => {
		testOptions.alias = `my-utils:${process.cwd()}/src/utils`;
		dependenciesAction(
			{ target: 'my-utils/cookiesServices.js' },
			testOptions
		).then(() => {
			expect(fs.writeFileSync.mock.calls).toHaveLength(1);
			const lastReportWriteCall = fs.writeFileSync.mock.calls[0];
			const [fileName, reportContents] = lastReportWriteCall;
			expect(reportContents).toMatchSnapshot();
			expect(openFile).lastCalledWith(fileName);
			expect(openFile).toHaveBeenCalledTimes(1);
			done();
		});
	});

	it('should not write a report when no results are found', done => {
		dependenciesAction({ target: 'src/noSuchMode.js' }, testOptions).then(
			() => {
				expect(fs.writeFileSync).not.toHaveBeenCalled();
				expect(openFile).not.toHaveBeenCalled();
				done();
			}
		);
	});
});
