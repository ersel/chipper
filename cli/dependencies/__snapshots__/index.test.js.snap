// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dependencies command should return all imports for an aliased search target 1`] = `
"
	<html>
	<head>
	<title>Chipper Scan Results</title>
	<style>
	
	table.blueTable {
		border: 1px solid #1C6EA4;
		background-color: #EEEEEE;
		width: 100%;
		text-align: left;
		border-collapse: collapse;
	  }
	  table.blueTable td, table.blueTable th {
		border: 1px solid #AAAAAA;
		padding: 3px 2px;
	  }
	  table.blueTable tbody td {
		font-size: 18px;
	  }
	  table.blueTable tr:nth-child(even) {
		background: #D0E4F5;
	  }
	  table.blueTable thead {
		background: #1C6EA4;
		background: -moz-linear-gradient(top, #5592bb 0%, #327cad 66%, #1C6EA4 100%);
		background: -webkit-linear-gradient(top, #5592bb 0%, #327cad 66%, #1C6EA4 100%);
		background: linear-gradient(to bottom, #5592bb 0%, #327cad 66%, #1C6EA4 100%);
		border-bottom: 2px solid #444444;
	  }
	  table.blueTable thead th {
		font-size: 15px;
		font-weight: bold;
		color: #FFFFFF;
		border-left: 2px solid #D0E4F5;
	  }
	  table.blueTable thead th:first-child {
		border-left: none;
	  }
	
	</style>
	</head>
	<table class=\\"blueTable\\">
	<thead>
	<tr>
	<th>Source Module</th><th>Imported Module</th><th>Imported Symbols</th></tr></thead><tr><td>/src/utils/cookiesServices.js</td><td>/node_modules/cookie-jar/index.js</td><td>setCookie, getCookie</td></tr><tr><td>/src/utils/cookiesServices.js</td><td>/node_modules/lodash/get/index.js</td><td>get</td></tr></table></html>"
`;

exports[`dependencies command should return all imports of a directory of modules 1`] = `
"
	<html>
	<head>
	<title>Chipper Scan Results</title>
	<style>
	
	table.blueTable {
		border: 1px solid #1C6EA4;
		background-color: #EEEEEE;
		width: 100%;
		text-align: left;
		border-collapse: collapse;
	  }
	  table.blueTable td, table.blueTable th {
		border: 1px solid #AAAAAA;
		padding: 3px 2px;
	  }
	  table.blueTable tbody td {
		font-size: 18px;
	  }
	  table.blueTable tr:nth-child(even) {
		background: #D0E4F5;
	  }
	  table.blueTable thead {
		background: #1C6EA4;
		background: -moz-linear-gradient(top, #5592bb 0%, #327cad 66%, #1C6EA4 100%);
		background: -webkit-linear-gradient(top, #5592bb 0%, #327cad 66%, #1C6EA4 100%);
		background: linear-gradient(to bottom, #5592bb 0%, #327cad 66%, #1C6EA4 100%);
		border-bottom: 2px solid #444444;
	  }
	  table.blueTable thead th {
		font-size: 15px;
		font-weight: bold;
		color: #FFFFFF;
		border-left: 2px solid #D0E4F5;
	  }
	  table.blueTable thead th:first-child {
		border-left: none;
	  }
	
	</style>
	</head>
	<table class=\\"blueTable\\">
	<thead>
	<tr>
	<th>Source Module</th><th>Imported Module</th><th>Imported Symbols</th></tr></thead><tr><td>/src/documentsService.js</td><td>/node_modules/lodash/lodash.js</td><td>uniq</td></tr><tr><td>/src/loginService.js</td><td>/src/documentsService.js</td><td>getDocuments as getLoggedInUsersDocuments</td></tr><tr><td>/src/loginService.js</td><td>/src/utils/cookiesService.js</td><td>setCookie</td></tr><tr><td>/src/utils/cookiesServices.js</td><td>/node_modules/cookie-jar/index.js</td><td>setCookie, getCookie</td></tr><tr><td>/src/utils/cookiesServices.js</td><td>/node_modules/lodash/get/index.js</td><td>get</td></tr></table></html>"
`;

exports[`dependencies command should return all imports of a single module (absolute path) 1`] = `
"
	<html>
	<head>
	<title>Chipper Scan Results</title>
	<style>
	
	table.blueTable {
		border: 1px solid #1C6EA4;
		background-color: #EEEEEE;
		width: 100%;
		text-align: left;
		border-collapse: collapse;
	  }
	  table.blueTable td, table.blueTable th {
		border: 1px solid #AAAAAA;
		padding: 3px 2px;
	  }
	  table.blueTable tbody td {
		font-size: 18px;
	  }
	  table.blueTable tr:nth-child(even) {
		background: #D0E4F5;
	  }
	  table.blueTable thead {
		background: #1C6EA4;
		background: -moz-linear-gradient(top, #5592bb 0%, #327cad 66%, #1C6EA4 100%);
		background: -webkit-linear-gradient(top, #5592bb 0%, #327cad 66%, #1C6EA4 100%);
		background: linear-gradient(to bottom, #5592bb 0%, #327cad 66%, #1C6EA4 100%);
		border-bottom: 2px solid #444444;
	  }
	  table.blueTable thead th {
		font-size: 15px;
		font-weight: bold;
		color: #FFFFFF;
		border-left: 2px solid #D0E4F5;
	  }
	  table.blueTable thead th:first-child {
		border-left: none;
	  }
	
	</style>
	</head>
	<table class=\\"blueTable\\">
	<thead>
	<tr>
	<th>Source Module</th><th>Imported Module</th><th>Imported Symbols</th></tr></thead><tr><td>/src/loginService.js</td><td>/src/documentsService.js</td><td>getDocuments as getLoggedInUsersDocuments</td></tr><tr><td>/src/loginService.js</td><td>/src/utils/cookiesService.js</td><td>setCookie</td></tr></table></html>"
`;

exports[`dependencies command should return all imports of a single module 1`] = `
"
	<html>
	<head>
	<title>Chipper Scan Results</title>
	<style>
	
	table.blueTable {
		border: 1px solid #1C6EA4;
		background-color: #EEEEEE;
		width: 100%;
		text-align: left;
		border-collapse: collapse;
	  }
	  table.blueTable td, table.blueTable th {
		border: 1px solid #AAAAAA;
		padding: 3px 2px;
	  }
	  table.blueTable tbody td {
		font-size: 18px;
	  }
	  table.blueTable tr:nth-child(even) {
		background: #D0E4F5;
	  }
	  table.blueTable thead {
		background: #1C6EA4;
		background: -moz-linear-gradient(top, #5592bb 0%, #327cad 66%, #1C6EA4 100%);
		background: -webkit-linear-gradient(top, #5592bb 0%, #327cad 66%, #1C6EA4 100%);
		background: linear-gradient(to bottom, #5592bb 0%, #327cad 66%, #1C6EA4 100%);
		border-bottom: 2px solid #444444;
	  }
	  table.blueTable thead th {
		font-size: 15px;
		font-weight: bold;
		color: #FFFFFF;
		border-left: 2px solid #D0E4F5;
	  }
	  table.blueTable thead th:first-child {
		border-left: none;
	  }
	
	</style>
	</head>
	<table class=\\"blueTable\\">
	<thead>
	<tr>
	<th>Source Module</th><th>Imported Module</th><th>Imported Symbols</th></tr></thead><tr><td>/src/loginService.js</td><td>/src/documentsService.js</td><td>getDocuments as getLoggedInUsersDocuments</td></tr><tr><td>/src/loginService.js</td><td>/src/utils/cookiesService.js</td><td>setCookie</td></tr></table></html>"
`;
